File: Lv 3/네트워크.py
def union(parent, a, b):
    t1 = find(parent, a)
    t2 = find(parent, b)
    
    if t1 != t2:
        parent[t1] = parent[t2]

def find(parent, a):
    if parent[a] != a:
        parent[a] = find(parent, parent[a])

    return parent[a]

def solution(n, computers):
    answer = 0
    parent = [i for i in range(n)]
    
    for i in range(n):
        for j in range(i+1, n):
            if computers[i][j] == 1:
                t1 = find(parent, i)
                t2 = find(parent, j)
                if t1 != t2:
                    union(parent, t1, t2)
    
    count = dict()
    for i in range(n):
        find(parent, i) # Tuning
    
    answer = len(set(parent))
    return answer

File: Lv 3/네트워크.java
import java.util.Arrays;

class Solution {

    public void union(int[] parent, int p1, int p2) {
        int parent1 = find(parent, p1);
        int parent2 = find(parent, p2);

        parent[parent1] = parent2;
    }

    public int find(int[] parent, int idx) {
        if (parent[idx] != idx) {
            parent[idx] = find(parent, parent[idx]);
        }

        return parent[idx];
    }

    public int solution(int n, int[][] computers) {
        int[] parent = new int[n];
        for (int i = 0; i < parent.length; i++) {
            parent[i] = i;
        }

        for (int i = 0; i < n; i++) {
            for (int j = i+1; j < n; j++) {
                if (computers[i][j] == 1) {
                    union(parent, i, j);
                }
            }
        }

        int answer = (int) Arrays.stream(parent)
                .map(i -> find(parent, i))
                .distinct()
                .count();

        return answer;
    }
}
